otx/mpa/modules/models/heads/semisl_cls_head.py

```diff
@@ -136,3 +138,140 @@ def forward_train(self, x, gt_label):
+        act_cfg (dict): Config of activation layer.
+        loss (dict): configuration of loss, default is CrossEntropyLoss
+        topk (set): evaluation topk score, default is (1, )
+        unlabeled_coef (float): unlabeled loss coefficient
+        dynamic_threshold (boolean): whether to use dynamic threshold, default is True
+        min_threshold (float): Minimum value of threshold determining pseudo-label, default is 0.5
+    """
+
+    def __init__(
+        self,
+        num_classes,
+        in_channels,
+        hid_channels=1280,
+        act_cfg=dict(type="ReLU"),
+        loss=dict(type="CrossEntropyLoss", loss_weight=1.0),
+        topk=(1,),
+        dropout=False,
+        unlabeled_coef=1.0,
+        use_dynamic_threshold=True,
+        min_threshold=0.5,
+    ):
+        if in_channels <= 0:
+            raise ValueError(f"in_channels={in_channels} must be a positive integer")
+        if num_classes <= 0:
+            raise ValueError("at least one class must be exist num_classes.")
+
+        topk = (1,) if num_classes < 5 else (1, 5)
+        super(SemiNonLinearClsHead, self).__init__(
+            num_classes, in_channels, hid_channels=hid_channels, act_cfg=act_cfg, loss=loss, topk=topk, dropout=dropout
+        )
+        self.unlabeled_coef = unlabeled_coef
+
+        # class wise accuracy for dynamic Threshold (min_thr ~ 1.0)
+        self.use_dynamic_threshold = use_dynamic_threshold
+        self.min_threshold = min_threshold if self.use_dynamic_threshold else 0.95
+        self.num_pseudo_label = 0
+        self.classwise_acc = torch.ones((self.num_classes,)) * self.min_threshold
+        if torch.cuda.is_available():
+            self.classwise_acc = self.classwise_acc.cuda()
+
+    def loss(self, logits, gt_label, pseudo_label=None, mask=None):
+        """loss function in which unlabeled data is considered
+
+        Args:
+            logit (set): (labeled data logit, unlabeled data logit)
+            gt_label (Tensor): target features for labeled data
+            pseudo_label (Tensor): target feature for unlabeled data
+            mask (Tensor): Mask that shows pseudo-label that passes threshold
+
+        Returns:
+            dict[str, Tensor]: A dictionary of loss components.
+        """
+        logits_x, logits_u_s = logits
+        num_samples = len(logits_x)
+        losses = dict()
+
+        # compute supervised loss
+        lx = self.compute_loss(logits_x, gt_label, avg_factor=num_samples)
+
+        lu = 0
+        if len(logits_u_s) > 0:
+            # compute unsupervised loss
+            lu = self.compute_loss(logits_u_s, pseudo_label, avg_factor=len(logits_u_s)) * mask
+        losses["loss"] = lx + self.unlabeled_coef * lu
+
+        # compute accuracy
+        acc = self.compute_accuracy(logits_x, gt_label)
+        losses["accuracy"] = {f"top-{k}": a for k, a in zip(self.topk, acc)}
+        return losses
+
+    def forward_train(self, x, gt_label):
```

Comment:
Is there a difference between the corresponding functions (forward_train, etc..) for Linear and Nonlinear? If not, this is considered duplicate code.

Reply:
Nonlinear uses `classifier` instead of `fc`. It should be separated. Also, inheriting linear head for non-linear head has been looked weird (I've tried but it showed poor readability).

Reply:
Then, how about creating a `SemiLinearHead` that has a separate `SemiSLHead` and inherits `SemiSLHead` and `LinearHead`? (NonLinear as well)

- SemiSLHead : Head for SemiSL algorithm
- SemiLinearHead: SemiSLHead + LinearHead
- SemiNonLinearHead: SemiSLHead + NonLinearHead

If you make it like this, the class will increase by one, but I think you can handle redundant code.

Reply:
I agree with harim, if head gets linear or non-linear flag, it can reduce lots of codes

Reply:
I'll refactor this file.

Reply:
Refactored. Please review again.